File: sdk/ruby/lib/cordum/cap/heartbeat.rb
# frozen_string_literal: true

module Cordum
  module Cap
    module HeartbeatHelper
      # Builds a basic heartbeat payload (serialized BusPacket bytes).
      def self.heartbeat_payload(worker_id:, pool:, active_jobs:, max_parallel:, cpu_load:)
        heartbeat_payload_with_progress(
          worker_id: worker_id, pool: pool,
          active_jobs: active_jobs, max_parallel: max_parallel,
          cpu_load: cpu_load
        )
      end

      # Builds a heartbeat payload with memory load.
      def self.heartbeat_payload_with_memory(worker_id:, pool:, active_jobs:, max_parallel:, cpu_load:, memory_load:)
        heartbeat_payload_with_progress(
          worker_id: worker_id, pool: pool,
          active_jobs: active_jobs, max_parallel: max_parallel,
          cpu_load: cpu_load, memory_load: memory_load
        )
      end

      # Builds a heartbeat payload with all fields including progress.
      def self.heartbeat_payload_with_progress(worker_id:, pool:, active_jobs:, max_parallel:,
                                                cpu_load:, memory_load: 0.0, progress_pct: 0, last_memo: '')
        hb = Cordum::Agent::V1::Heartbeat.new(
          worker_id: worker_id,
          pool: pool,
          active_jobs: active_jobs,
          max_parallel_jobs: max_parallel,
          cpu_load: cpu_load.to_f,
          memory_load: memory_load.to_f,
          progress_pct: progress_pct,
          last_memo: last_memo
        )

        pkt = Cordum::Agent::V1::BusPacket.new(
          sender_id: worker_id,
          protocol_version: DEFAULT_PROTOCOL_VERSION,
          created_at: Google::Protobuf::Timestamp.new(seconds: Time.now.to_i, nanos: 0),
          heartbeat: hb
        )

        Codec.marshal_deterministic(pkt)
      end

      # Publishes a heartbeat payload to NATS.
      def self.emit(nc, payload)
        nc.publish(SUBJECT_HEARTBEAT, payload)
      end
    end

    # Background thread that periodically emits heartbeats.
    class HeartbeatLoop
      # @param nc [NATS::IO::Client] NATS connection
      # @param payload_fn [Proc] callable returning heartbeat payload bytes
      # @param interval [Numeric] seconds between heartbeats
      # @param metrics [MetricsHook, nil] optional metrics hook
      # @param worker_id [String] worker ID for metrics
      def initialize(nc:, payload_fn:, interval: DEFAULT_HEARTBEAT_INTERVAL, metrics: nil, worker_id: '')
        @nc = nc
        @payload_fn = payload_fn
        @interval = interval
        @metrics = metrics
        @worker_id = worker_id
        @running = false
        @thread = nil
      end

      # Starts the heartbeat loop in a background thread.
      def start
        @running = true
        @thread = Thread.new do
          while @running
            tick
            sleep @interval
          end
        end
        @thread.abort_on_exception = true
        self
      end

      # Executes a single heartbeat tick. Useful for external event loops.
      def tick
        payload = @payload_fn.call
        HeartbeatHelper.emit(@nc, payload)
        @metrics&.on_heartbeat_sent(@worker_id)
      rescue StandardError => e
        @metrics&.on_error('heartbeat', e.message)
      end

      # Stops the heartbeat loop and waits for the thread to finish.
      def stop
        @running = false
        @thread&.join(2)
      end

      def running?
        @running
      end
    end
  end
end

File: sdk/dotnet/src/Cordum.Cap/Heartbeat.cs
using Google.Protobuf.WellKnownTypes;
using NATS.Client.Core;
using Cordum.Agent.V1;

namespace Cordum.Cap;

/// <summary>
/// Heartbeat payload constructors and emission helpers.
/// </summary>
public static class HeartbeatHelper
{
    /// <summary>
    /// Build a heartbeat payload with basic fields.
    /// </summary>
    public static byte[] HeartbeatPayload(
        string workerId, string pool, int activeJobs, int maxParallel, float cpuLoad)
    {
        return HeartbeatPayloadWithProgress(workerId, pool, activeJobs, maxParallel, cpuLoad, 0f, 0, "");
    }

    /// <summary>
    /// Build a heartbeat payload with memory load.
    /// </summary>
    public static byte[] HeartbeatPayloadWithMemory(
        string workerId, string pool, int activeJobs, int maxParallel, float cpuLoad, float memoryLoad)
    {
        return HeartbeatPayloadWithProgress(workerId, pool, activeJobs, maxParallel, cpuLoad, memoryLoad, 0, "");
    }

    /// <summary>
    /// Build a heartbeat payload with all fields including progress and memo.
    /// </summary>
    public static byte[] HeartbeatPayloadWithProgress(
        string workerId, string pool, int activeJobs, int maxParallel,
        float cpuLoad, float memoryLoad, int progressPct, string lastMemo)
    {
        var hb = new Heartbeat
        {
            WorkerId = workerId,
            Pool = pool,
            ActiveJobs = activeJobs,
            MaxParallelJobs = maxParallel,
            CpuLoad = cpuLoad,
            MemoryLoad = memoryLoad,
            ProgressPct = progressPct,
            LastMemo = lastMemo,
        };

        var packet = new BusPacket
        {
            SenderId = workerId,
            ProtocolVersion = Subjects.DefaultProtocolVersion,
            CreatedAt = Timestamp.FromDateTimeOffset(DateTimeOffset.UtcNow),
            Heartbeat = hb,
        };

        return Codec.MarshalDeterministic(packet);
    }

    /// <summary>
    /// Publish a heartbeat payload to the heartbeat subject.
    /// </summary>
    public static async Task EmitHeartbeat(NatsConnection nc, byte[] payload, CancellationToken ct = default)
    {
        await nc.PublishAsync(Subjects.Heartbeat, payload, cancellationToken: ct);
    }
}

/// <summary>
/// Periodically emits heartbeats using a PeriodicTimer.
/// </summary>
public sealed class HeartbeatLoop : IAsyncDisposable
{
    private readonly NatsConnection _nc;
    private readonly Func<byte[]> _payloadFn;
    private readonly string _workerId;
    private readonly TimeSpan _interval;
    private readonly IMetricsHook _metrics;
    private readonly CancellationTokenSource _cts = new();
    private Task? _loopTask;

    public HeartbeatLoop(
        NatsConnection nc,
        Func<byte[]> payloadFn,
        string workerId,
        TimeSpan interval,
        IMetricsHook? metrics = null)
    {
        _nc = nc;
        _payloadFn = payloadFn;
        _workerId = workerId;
        _interval = interval;
        _metrics = metrics ?? new NoopMetrics();
    }

    /// <summary>
    /// Start the heartbeat loop. Returns immediately; heartbeats run in the background.
    /// </summary>
    public void Start()
    {
        _loopTask = RunLoop(_cts.Token);
    }

    private async Task RunLoop(CancellationToken ct)
    {
        using var timer = new PeriodicTimer(_interval);
        try
        {
            while (await timer.WaitForNextTickAsync(ct))
            {
                try
                {
                    var payload = _payloadFn();
                    await HeartbeatHelper.EmitHeartbeat(_nc, payload, ct);
                    _metrics.OnHeartbeatSent(_workerId);
                }
                catch (OperationCanceledException)
                {
                    break;
                }
                catch
                {
                    // Swallow individual emit failures; loop continues
                }
            }
        }
        catch (OperationCanceledException)
        {
            // Normal shutdown
        }
    }

    /// <summary>
    /// Stop the heartbeat loop and wait for it to finish.
    /// </summary>
    public async Task StopAsync()
    {
        await _cts.CancelAsync();
        if (_loopTask != null)
        {
            await _loopTask;
        }
    }

    public async ValueTask DisposeAsync()
    {
        await StopAsync();
        _cts.Dispose();
    }
}
